lib/utils/bcrypt.go

```diff
@@ -0,0 +1,39 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"golang.org/x/crypto/bcrypt"
+)
+
+const maxInputSize = 72
+
+// truncateToMaxSize Make sure input is truncated to the maximum length crypto accepts.  Crypto changed the behavior
+// from ignoring the extra input to returning an error, this truncation is necessary to maintain compatibility with
+// customers who have long passwords, or more commonly our recovery codes.
+func truncateToMaxSize(input []byte) []byte {
+	if len(input) > maxInputSize {
+		return input[:maxInputSize]
+	}
```

Comment:
Something like `input[:subtle.ConstantTimeSelect(subtle.ConstantTimeLessOrEq(len(input), 72), len(input), 72)]`

Reply:
This PR is not attempting to pretend that additional input is calculated as part of the hash.  So using a timing attack to determine that the password is truncated does not provide any value (the truncation is not supposed to be a secret process).

Let me know if you see another risk I may not have considered.  Thank you @espadolini!

Reply:
I suppose the only thing you'd learn is that the password was intended to be longer than 72 and is now 72, versus the password being between 1 and 71 characters, which doesn't really help.

Reply:
And as an attacker that information does not help you, since you would only need to target the first 72 characters anyways

Reply:
Well, correction, it would help in knowing that you would need to target exactly 72 characters.  Although brute force of a password of that length is infeasible.